primitives/arithmetic/src/per_things.rs

```diff
@@ -30,29 +30,62 @@ pub trait PerThing: Sized + Saturating + Copy {
 	/// The data type used to build this per-thingy.
 	type Inner: BaseArithmetic + Copy;
 
-	/// accuracy of this type
-	const ACCURACY: Self::Inner;
+	/// A data type larger than `Self::Inner`, used to avoid overflow in some computations.
+	/// It must be able to compute `ACCURACY^2`.
+	type Upper: BaseArithmetic + Copy + From<Self::Inner> + TryInto<Self::Inner>;
 
-	/// NoThing
-	fn zero() -> Self;
+	/// The accuracy of this type.
+	const ACCURACY: Self::Inner;
 
-	/// `true` if this is nothing.
-	fn is_zero(&self) -> bool;
+	/// Equivalent to `Self::from_parts(0)`.
+	fn zero() -> Self { Self::from_parts(Self::Inner::zero()) }
 
-	/// Everything.
-	fn one() -> Self;
+	/// Return `true` if this is nothing.
+	fn is_zero(&self) -> bool { self.deconstruct() == Self::Inner::zero() }
 
-	/// Consume self and deconstruct into a raw numeric type.
-	fn deconstruct(self) -> Self::Inner;
+	/// Equivalent to `Self::from_parts(Self::ACCURACY)`.
+	fn one() -> Self { Self::from_parts(Self::ACCURACY) }
 
-	/// From an explicitly defined number of parts per maximum of the type.
-	fn from_parts(parts: Self::Inner) -> Self;
+	/// Return `true` if this is one.
+	fn is_one(&self) -> bool { self.deconstruct() == Self::ACCURACY }
 
-	/// Converts a percent into `Self`. Equal to `x / 100`.
-	fn from_percent(x: Self::Inner) -> Self;
+	/// Build this type from a percent. Equivalent to `Self::from_parts(x * Self::ACCURACY / 100)`.
+	fn from_percent(x: Self::Inner) -> Self {
+		Self::from_parts(x.min(100.into()) * (Self::ACCURACY / 100.into()))
+	}
 
 	/// Return the product of multiplication of this value by itself.
-	fn square(self) -> Self;
+	fn square(self) -> Self {
+		let p = Self::Upper::from(self.deconstruct());
+		let q = Self::Upper::from(Self::ACCURACY); 
+		Self::from_rational_approximation(p * p, q * q)
+	}
+
+	/// Saturating reciprocal multiplication. Compute `x / self`, saturating at the numeric 
+	/// bounds instead of overflowing.
+	fn saturating_reciprocal_mul<N>(self, x: N) -> N
+	where N: UniqueSaturatedInto<Self::Inner> + UniqueSaturatedFrom<Self::Upper> {
+		if self.deconstruct() == 0.into() {
+			N::unique_saturated_from(Self::Upper::max_value())
+		} else {
+			let x = x.saturated_into::<Self::Inner>();
+			N::unique_saturated_from(Self::Upper::from(x) * Self::Upper::from(Self::ACCURACY) / Self::Upper::from(self.deconstruct()))
+		}
+	}
+
+	/// Saturating truncating multiplication. Compute `x * self`, truncating any fractional
+	/// remainder, IE rounding down to the nearest integer. 
+	fn saturating_truncating_mul<N>(self, x: N) -> N 
+	where N: UniqueSaturatedInto<Self::Inner> + UniqueSaturatedFrom<Self::Upper> {
+		let x = x.saturated_into::<Self::Inner>();
+		N::unique_saturated_from(Self::Upper::from(x) * Self::Upper::from(self.deconstruct()) / Self::Upper::from(Self::ACCURACY))
+	}
```

Comment:
here as well I think `Into<Self::Inner>` would be better, also this is a fast subset implementation of trait `Mul` maybe we should put in the comment that for other usage one could use `Mul` instead.